### PR TITLE
Feature/pyrb

### LINKED ISSUE #1675: Sampling [log-]randomly
To get a sample for a problem, I created a function
```python
def listOfParams(n):
    mus = []
    for _ in range(n):
        mus.append(Dmu.element(True, True))    # the second True stand for 'log-random'
    return mus
```
(maybe there already exists a function generating such a sampling)

For a problem where there is only 1 parameter living in _[0.1, 10]_ I created two samples, one with the _log-random_ option, and another without.

Here is what I get when I try to compare the two samples :
image
and in log-scale :
image
When I run the greedy algorithm on a more complex case where parameters have a range with many orders of magnitude, there is no significant difference between the results given by the log-random generated sample and the other one.

The discussion is about how samples are generated.

### LINKED ISSUE #1732: set [automatic] tests to check the code of reduced basis

### LINKED ISSUE #1768: Set tests for pyfeelpp-mor
Set tests to test the module `feelpp.mor.reducedbasis` (stationnary and time-dependant) to be run in clean environment using pytest and continuous integration

### LINKED ISSUE #1888: Reduced Basis: Wrong output computation in online step
I work in the `feature/pyrb` branch.
I try to use the Python script to generate reduced basis with the Toolbox Mor (in `feelpp / python / pyfeelpp-mor` folder), with the Greedy algorithm ( class `reducedbasisOffline` in `reducedbasisOffline.py` ).
I have the following warning:
```
[reducedBasis] Greedy, step 1:[0%                                                                        ]
reducedbasis.py:351: RuntimeWarning: invalid value encountered in sqrt
return normHatE / np.sqrt(alp) 
```
The function mentioned is:
https://github.com/feelpp/feelpp/blob/db39f6ef1c7182f6ed6c3ec043265c39e7473abb/python/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py#L334-L351
However the algorithm does not stop and so when an output is computed in the online step, the result is totally wrong.
I pushed my testcases (2d and 3d) with the commit 3ae667be63ddd7a109692d7048b5c7d34401867f.

### PR DESCRIPTION
**Pull request by thomas-saigre**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [x] Have you written Doxygen comments in your contribution ?

-----

- closes #1732
- closes #1675
- implement reduced basis method in python using feelpp and petsc4py
- closes #1768
- closes #1888
- #1883

This PR contains the module reducedbasis use the run the offline and online part for the reduced basis method.
For now, only steady-time cases can be treated, and only "coercive" problem, as long as the scm module has not yet been wrapped in python, so it uses the min-theta approach for the computation of error bound.
This is not pushed yet, but as soon I have tested it will be, and different outputs can be studied (for the not-compliant case).

The content has two main parts :

The python module reducedbasis contains all the classes and code for offline and online parts
The script generate_basis.py allows to generate a reduced basis using one of 3 algorithms (from a fixed random sampling, from a POD construction, of the greedy algorithm)
There are still some issues in the tests pipeline due to persistent files not compatible between 2d and 3d cases #1768, which I did not manage to resolve, but when I run by hand all the tests locally, deleting by hand the directory feel/crbdb and `feel/pyfeelppmor-tests, the tests pass well both in sequential and parallel.

### CODE REVIEW

**Comment by prudhomm** [comment]
@thomas-saigre tests are not passing: https://buildkite.com/feelpp/feelpp/builds/4189#e4c4a737-2160-44d3-9c64-c61ec241110a/749-1132

**Comment by thomas-saigre** [comment]
Yes, according to log, the python package `tqdm` is not installed on the computer rinnong the tests (maybe that is not the only problem)

**Comment by prudhomm** [comment]
@thomas-saigre I don't see the mention of tqdm. tqdm is installed, the crash is elsewhere

**Comment by thomas-saigre** [comment]
yes saw it yesterday, but I perhaps was on the wrong page... I will look into the current crash that seems to occur on initialisation of the `heat` toolbox

File: mor/pyfeelpp-mor/tests/CMakeLists.txt [hunk 1]
```diff
@@ -1,22 +1,43 @@
...
-          mpirun -np 2 ${PYTHON_EXECUTABLE} -m pytest -k ${test} -s ${CMAKE_CURRENT_SOURCE_DIR}
-          WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-  )
-endforeach()
\ No newline at end of file
+add_test (NAME feelpp_mor_python-tests-mor
+COMMAND ${CMAKE_COMMAND} -E env
+      LD_LIBRARY_PATH=${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}:$ENV{LD_LIBRARY_PATH}
+      PYTHONPATH=${CMAKE_INSTALL_PREFIX}/${FEELPP_PYTHON_MODULE_PATH}
+      ${PYTHON_EXECUTABLE} -m pytest test_mor.py -s -vv
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:11]: why put everything in a single test file?

File: mor/pyfeelpp-mor/tests/CMakeLists.txt [hunk 2]
```diff
@@ -1,22 +1,43 @@
...
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
+add_test (NAME feelpp_mor_python-tests-mpi-mor
+COMMAND ${CMAKE_COMMAND} -E env
+      LD_LIBRARY_PATH=${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}:$ENV{LD_LIBRARY_PATH}
+      PYTHONPATH=${CMAKE_INSTALL_PREFIX}/${FEELPP_PYTHON_MODULE_PATH}
+      mpirun -np 2 ${PYTHON_EXECUTABLE} -m pytest test_mor.py -s -vv
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:19]: idem

File: mor/pyfeelpp-mor/tests/test_reducedbasis.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+FEEL_PATH = os.environ['HOME'] + '/feel'
+
+from mpi4py import MPI
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+
+
+from feelpp.toolboxes.heat import *
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/test_reducedbasis.py:15]: this depends on the heat toolbox, it should be reflected in the filename probably

File: mor/pyfeelpp-mor/tests/CMakeLists.txt (hunk 1 above)
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:11]: why put everything in the same file ?
with -k we can split

File: mor/pyfeelpp-mor/tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,22 +1,43 @@
...
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:20]: idem

**Review by prudhomm — APPROVED** [review]

**Comment by prudhomm** [comment]
@thomas-saigre some errors here https://buildkite.com/feelpp/feelpp/builds/4217#fb18ac13-b83d-4bc2-a21a-03243f055ad8/759-769

**Comment by thomas-saigre** [comment]
indeed... It was because vscode did not show whitespace diff in git diff, so I missed it when I committed...
To avoid it, I added `"diffEditor.ignoreTrimWhitespace": true,` in `seddtings.json`

**Comment by thomas-saigre** [comment]
@prudhomm Some tests didn't pass because `h5py` is not installed : https://buildkite.com/feelpp/feelpp/builds/4224#c813c6cc-7755-49aa-8e2f-a3ec1b870543/752-1255

**Comment by prudhomm** [comment]
the packages are not installed automatically, please discuss new packages installation before using them.
add an issue on repo feelpp/docker to indicate the debian package to install `python3-...` or pip if not available

also if you changed only mor then use 
`skip feelpp skip toolboxes` in the commit message to accelerate the build and avoid compiling feelpp and the toolboxes

**Comment by prudhomm** [comment]
@thomas-saigre there are no detailed description of the PR in issues right ?
could you improve the PR description please ?

**Comment by thomas-saigre** [comment]
This PR contains the module `reducedbasis` use the run the offline and online part for the reduced basis method.
For now, only steady-time cases can be treated, and only "coercive" problem, as long as the scm module has not yet been wrapped in python, so it uses the _min-theta_ approach for the computation of error bound.
This is not pushed yet, but as soon I have tested it will be, and different outputs can be studied (for the not-compliant case).

The content has two main parts :
- The python module `reducedbasis` contains all the classes and code for offline and online parts
- The script `generate_basis.py` allows to generate a reduced basis using one of 3 algorithms (from a fixed random sampling, from a POD construction, of the greedy algorithm)

There are still some issues in the tests pipeline due to persistent files not compatible between 2d and 3d cases, which I did not manage to resolve, but when I run by hand all the tests locally, deleting by hand the directory `feel/crbdb` and `feel/pyfeelppmor-tests, the tests pass well both in sequential and parallel.

**Comment by prudhomm** [comment]
@thomas-saigre do you have a list of issues that this PR closes ?

**Comment by prudhomm** [comment]
@thomas-saigre you have conflicts with the `develop` branch

**Comment by thomas-saigre** [comment]
> @thomas-saigre do you have a list of issues that this PR closes ?

* #1675 that is quite old, actually I don't remember the problem of it, this may be closed I think
* #1768 about tests, the tests are written, even if they don't pass the pipeline yet. I have merged `develop` in this branch, and I will push to run the pipeline from begining
* #1888 output still give weird values in python only (in c++, it works well), see the results of [test_eim.py](https://github.com/feelpp/feelpp/blob/feature/pyrb/python/pyfeelpp-mor/tests/test_eim.py)
* #1883 about online efficiency

**Comment by prudhomm** [comment]
Could you update the pr description with the closes keyword?

File: python/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis_timeOffline.py [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+
+        while Delta_max > eps_tol:
+
+            SN.append(mu_star)
+
+            # POD(t) step
+            if self.N == 0:
+                POD = self.computePODMode(mu_star, g, R=R, delta=delta, proj=False)
+                for ksi in POD:
+                    self.Z.append(ksi)
```
Comment (thomas-saigre) [inline python/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis_timeOffline.py:454]: there is still an issue when several POD modes are computed : the norm of the residual is not correct when the scalar product terms computed offline are used

Comment (prudhomm) [reply python/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis_timeOffline.py:454]: Weird. You remove the projection before applying the pod

**Pull request closed without merging**